File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.cpp
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        
        ListNode* node = head;
        for (int i = 0; i < k; ++i) {
            if (!node) return head;
            node = node->next;
        }

     
        ListNode* prev = nullptr;
        ListNode* curr = head;
        ListNode* next = nullptr;
        int count = 0;
        while (count < k) {
            next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
            count++;
        }

        head->next = reverseKGroup(curr, k);
        return prev;
    }
};

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        
         ListNode node = head;
        for (int i = 0; i < k; ++i) {
            if (node == null) return head;
            node = node.next;
        }

     
        ListNode  prev = null;
        ListNode  curr = head;
        ListNode  next = null;
        int count = 0;
        while (count < k) {
            next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
            count++;
        }

        head.next = reverseKGroup(curr, k);
        return prev;
    }
}